01-messaging.md

```diff
@@ -82,6 +83,50 @@ however, adding a 6-byte padding after the type field was considered
 wasteful: alignment may be achieved by decrypting the message into
 a buffer with 6-bytes of pre-padding.
 
+## Type-Length-Value Format
+
+In various places in the protocol we use a `tlv` format.  This is
+always of form:
+
+* [`var_int`:`type`]
+* [`var_int`:`length`]
+* [`length`:`value`]
+
+The `type` byte is never 0 (in some circumstances this can indicate the
+termination of the `tlv` stream).  The `type` and `length` fields use the bitcoin
+CompactSize encoding: fields less than 253 are single-byte, 253 is followed by a
+two-byte field, and 254 is followed by a four-byte field.  255 is
+invalid.  The `value` depends entirely on the `type` and context, though we
+explicitly allow `length` to be reduced for small integers.
+
+### Requirements
+
+The sending node:
+  - MUST NOT set `type` to 0.
+  - MUST use the shortest possible representation for `type` and `length`
+  - MUST order `tlv` in ascending type, length and value.
+  - if the `type` indicates that `value` is an integer:
+    - MUST NOT set `length` greater than 8.
+	- MAY truncate leading zeroes from `value` by shortening `length`.
+
+The receiving node:
+  - if `type` is unknown:
+	- if `type` is odd:
+		- MUST ignore the `tlv`
+	- otherwise `type` is even:
+		- MUST fail to parse the `tlv`
+  - otherwise, if `type` indicates that `value` is an integer:
+    - MUST handle `length` between 0 and 8 inclusive by prepending zero bytes.
+
+### Rationale
+
+This format allows optional information in a backwards compatible manner
+while still allowing non-backwards compatible changes in future.
+
+The requirement on writers for ordering ensures a canonical ordering
+for any `tlv` stream; readers do not check this for simplicity and
```

Comment:
What do we gain by not enforcing canonical ordering from the start? Apart from saving a few lines of code, to me this just invites more bugs (through side effects) in parsing duplicate records, or forcing the receiver to do more work in parsing the same field multiple times. It also means we can’t make other optimizations in decoding because we must support parsing the fields in any order. Simplicity and flexibility aren’t very compelling IMO, are there other reasons?